ci/rtools/awssdk_ep.patch

```diff
@@ -0,0 +1,213 @@
+@@ -134,7 +134,7 @@ namespace Aws
+                 {
+                     m_data.reset(Aws::NewArray<T>(m_size, ARRAY_ALLOCATION_TAG));
+
+-#ifdef _WIN32
++#ifdef _MSC_VER
+                     std::copy(other.m_data.get(), other.m_data.get() + other.m_size, stdext::checked_array_iterator< T * >(m_data.get(), m_size));
+ #else
+                     std::copy(other.m_data.get(), other.m_data.get() + other.m_size, m_data.get());
+diff --git a/aws-cpp-sdk-core/include/aws/core/utils/crypto/bcrypt/CryptoImpl.h b/aws-cpp-sdk-core/include/aws/core/utils/crypto/bcrypt/CryptoImpl.h
+index e26e36b60..3e7189b70 100644
+--- a/aws-cpp-sdk-core/include/aws/core/utils/crypto/bcrypt/CryptoImpl.h
++++ b/aws-cpp-sdk-core/include/aws/core/utils/crypto/bcrypt/CryptoImpl.h
+@@ -29,7 +29,14 @@ namespace Aws
+     {
+         namespace Crypto
+         {
+-            static const char* SecureRandom_BCrypt_Tag = "SecureRandom_BCrypt";
++#ifdef __MINGW32__
++#pragma GCC diagnostic push
++#pragma GCC diagnostic ignored "-Wunused-variable"
++#endif
++           static const char* SecureRandom_BCrypt_Tag = "SecureRandom_BCrypt";
++#ifdef __MINGW32__
++#pragma GCC diagnostic pop
++#endif
+
+             class SecureRandomBytes_BCrypt : public SecureRandomBytes
+             {
+diff --git a/aws-cpp-sdk-core/source/http/windows/WinHttpSyncHttpClient.cpp b/aws-cpp-sdk-core/source/http/windows/WinHttpSyncHttpClient.cpp
+index 4dade6489..a0456cf8e 100644
+--- a/aws-cpp-sdk-core/source/http/windows/WinHttpSyncHttpClient.cpp
++++ b/aws-cpp-sdk-core/source/http/windows/WinHttpSyncHttpClient.cpp
+@@ -22,6 +22,16 @@
+ #include <sstream>
+ #include <iostream>
+
++#ifndef WINHTTP_OPTION_WEB_SOCKET_KEEPALIVE_INTERVAL
++#define WINHTTP_OPTION_WEB_SOCKET_KEEPALIVE_INTERVAL 116
++#endif
++#ifndef WINHTTP_FLAG_SECURE_PROTOCOL_TLS1_1
++#define WINHTTP_FLAG_SECURE_PROTOCOL_TLS1_1 0x00000200
++#endif
++#ifndef WINHTTP_FLAG_SECURE_PROTOCOL_TLS1_2
++#define WINHTTP_FLAG_SECURE_PROTOCOL_TLS1_2 0x00000800
++#endif
++
+ using namespace Aws::Client;
+ using namespace Aws::Http;
+ using namespace Aws::Http::Standard;
+@@ -272,7 +282,7 @@ bool WinHttpSyncHttpClient::DoQueryHeaders(void* hHttpRequest, std::shared_ptr<H
+     wmemset(contentTypeStr, 0, static_cast<size_t>(dwSize / sizeof(wchar_t)));
+
+     WinHttpQueryHeaders(hHttpRequest, WINHTTP_QUERY_CONTENT_TYPE, nullptr, &contentTypeStr, &dwSize, 0);
+-    if (contentTypeStr[0] != NULL)
++    if (contentTypeStr[0])
+     {
+         Aws::String contentStr = StringUtils::FromWString(contentTypeStr);
+         response->SetContentType(contentStr);
+@@ -303,7 +313,7 @@ bool WinHttpSyncHttpClient::DoQueryHeaders(void* hHttpRequest, std::shared_ptr<H
+
+ bool WinHttpSyncHttpClient::DoSendRequest(void* hHttpRequest) const
+ {
+-    return (WinHttpSendRequest(hHttpRequest, NULL, NULL, 0, 0, 0, NULL) != 0);
++    return (WinHttpSendRequest(hHttpRequest, NULL, 0, 0, 0, 0, 0) != 0);
```

Comment:
Let's use NULL for all [pointer arguments](https://learn.microsoft.com/en-us/windows/win32/api/winhttp/nf-winhttp-winhttpsendrequest)?
```suggestion
-    return (WinHttpSendRequest(hHttpRequest, NULL, NULL, 0, 0, 0, NULL) != 0);
+    return (WinHttpSendRequest(hHttpRequest, NULL, 0, NULL, 0, 0, NULL) != 0);
```

Reply:
(yes, this is a diff of diff of diff)

Reply:
Sorry I missed this comment. Fixed locally and testing now. I'll update here when this is fixed.

Reply:
Hrm, this errors,

```
C:/Users/Bryce/Source/apache/arrow/src/build-i686-cpp/awssdk_ep-prefix/src/awssdk_ep/aws-cpp-sdk-core/source/http/windows/WinHttpSyncHttpClient.cpp:316:71: error: passing NULL to non-pointer argument 7 of 'WINBOOL WinHttpSendRequest(HINTERNET, LPCWSTR, DWORD, LPVOID, DWORD, DWORD, DWORD_PTR)' [-Werror=conversion-null]
     return (WinHttpSendRequest(hHttpRequest, NULL, 0, NULL, 0, 0, NULL) != 0);
```

Reply:
Change made in 19cf666a70350a307bd745dbb812f4fee828f756. Tested locally that it works.